src/rmarkdown/preview.ts

```diff
@@ -0,0 +1,179 @@
+
+import * as cp from 'child_process';
+import * as kill from 'tree-kill';
+import * as vscode from 'vscode';
+import { getBrowserHtml } from '../session';
+import { closeBrowser, isHost, shareBrowser } from '../liveshare';
+import { config, doWithProgress, getRpath,  setContext } from '../util';
+
+interface IPreviewProcess {
+    cp: cp.ChildProcessWithoutNullStreams,
+    file: string,
+    panel: vscode.WebviewPanel,
+}
+
+export class RMarkdownPreviewManager {
+    private openProcesses: IPreviewProcess[] = [];
+    private activePreview: vscode.WebviewPanel;
+    private activeResource: vscode.Uri;
+    private activeExternalResource: vscode.Uri;
+    private rPath: string;
+
+    public async init(): Promise<void> {
+        this.rPath = await getRpath(false);
+    }
+
+    public async previewRmd(viewer: vscode.ViewColumn, uri?: vscode.Uri): Promise<void> {
+        const fileUri = uri ?? vscode.window.activeTextEditor.document.uri;
+        const fileName = fileUri.path.substring(fileUri.path.lastIndexOf('/') + 1);
+        const previewEngine: string = config().get('rmarkdown.previewEngine');
+        const cmd = (
+            `${this.rPath} --silent --slave --no-save --no-restore -e "${previewEngine}('${fileUri.path}')"`
+        );
+        const reg: RegExp = this.constructRegex(previewEngine);
+        let call = undefined;
+
+
+        if (this.openProcesses.some(e => e.file === fileName)) {
+            this.openProcesses.filter(e => e.file === fileName)[0].panel.reveal();
+        } else {
+            await doWithProgress(() => {
+                try {
+                    call = cp.spawn(cmd, null, { shell: true });
+                } catch (e) {
+                    console.warn((e as string));
+                }
+                (call as cp.ChildProcessWithoutNullStreams).stderr.on('data',
+                    (data: Buffer) => {
+                        const dat = data.toString('utf8');
+                        const match = reg.exec(dat)?.[0];
+                        const previewUrl = this.constructUrl(previewEngine, match, fileName);
+                        if (match) {
+                            void this.showPreview(previewUrl, fileName, call, viewer, fileUri);
+                        }
+                    });
+            },
+                vscode.ProgressLocation.Notification,
+                `Rendering ${fileName}...`
+            );
+        }
+    }
+
+    public refreshPanel(): void {
+        if (this.activePreview) {
+            this.activePreview.webview.html = '';
+            this.activePreview.webview.html = getBrowserHtml(this.activeExternalResource);
+        }
+    }
+
+    public async showSource(): Promise<void> {
+        if (this.activeResource) {
+            const viewCol = vscode.window.visibleTextEditors.filter(e => e.document.uri === this.activeResource)[0]?.viewColumn;
```

Comment:
This seems not working as expected for me. When the source is on the first view column but not selected and the preview on the second view column, then `showSource()` will open a new tab in the second view column (seems to fallback to `Active`) so that the original source editor in the first view column is not revealed.

Reply:
I could replicate this, but only if I did the following:

1. Open rmd file
2. Preview doc
3. Close rmd file
4. Reopen rmd file
5. Go to source

I'll dig into this a bit